typescript/patterns/src/xa_transactions.ts

```diff
@@ -0,0 +1,123 @@
+import * as restate from "@restatedev/restate-sdk";
+import { randomUUID } from "crypto";
+
+//  -----------------                                         -----------------
+//                      Coordinating Database transactions
+//  -----------------                                         -----------------
+
+// Restate can be used to be the transaction coordinator for a database
+// transaction. The query to the database will be committed exactly once, and can
+// thus participate consistently in a sequence of steps in a workflow or durable
+// handler execution. 
+//
+// To do that, we use the 2-phase-commit protocol approach, implemented by many
+// databases via the XA transaction standard. For example, PostgreSQL supports this
+// via 'prepared transactions': https://www.postgresql.org/docs/current/sql-prepare-transaction.html
+//
+// Unlike the classical 2-phase-commit scenario, we are not using the protocol to
+// commit according to vote 
+
+// In principle, the operation is as simple as
+// (1) sideEffect: (exequte query, prepare transaction under ID, return ID from sideEffect).
+//                 This side effect commits once, to a single ID is stored in Restate for
+//                 committing.
+// (2) sideEffect: commit the prepared transaction
+//
+// The slightly higher complexity here comes from the fact that we need to ensure that
+// any transaction that was prepared, but will not be committed, must be cleaned up by
+// the recovery run. Otherwise that transaction will keep holding database locks and
+// block other transactions. We also re-try the transaction after that.
+
+// ----------------------------------------------------------------------------
```

Comment:
We should also highlight the imperative to minimize the time between prepare and commit steps. Maybe we can make that explicit by putting an explicit timeout into the pattern? But, it would make the example even harder to follow.